Sources/Fuzzilli/FuzzIL/ProgramTypes.swift

```diff
@@ -0,0 +1,90 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// Store known types for program variables at specific instructions
+public struct TypeIndexPair: Equatable {
+    public let index: Int
+    public let type: Type
+
+    public init(index: Int, type: Type){
+        self.index = index
+        self.type = type
+    }
+
+    public static func == (lhs: TypeIndexPair, rhs: TypeIndexPair) -> Bool {
+        return lhs.index == rhs.index && lhs.type == rhs.type
+    }
+}
+
+public struct ProgramTypes: Equatable, Sequence {
+    // types[variable][instructionIndex] = type
+    private var types = VariableMap<[TypeIndexPair]>()
+
+    public init () {}
+
+    public init (from types: VariableMap<[TypeIndexPair]>) {
+        self.types = types
+    }
+
+    public static func == (lhs: ProgramTypes, rhs: ProgramTypes) -> Bool {
+        return lhs.types == rhs.types
+    }
+
+    // Save type of given variable
+    public mutating func setType(of variable: Variable, to type: Type, at instrIndex: Int) {
+        // Initialize type structure for given variable if not already
+        if types[variable] == nil {
+            types[variable] = []
+        }
+        let typeIndexPair = TypeIndexPair(index: instrIndex, type: type)
+        // Check if we update type of instruction in the middle
+        if let insertionIndex = types[variable]!.firstIndex(where: { el in el.index >= instrIndex }) {
+            if types[variable]![insertionIndex].index == instrIndex {
+                // Overwrite old type
+                types[variable]![insertionIndex] = typeIndexPair
+            } else {
+                // Add new type information at instrIndex
+                types[variable]!.insert(typeIndexPair, at: insertionIndex)
+            }
+        } else {
+            types[variable]!.append(typeIndexPair)
+        }
+    }
+
+    // Get type of variable at current instruction
+    public func getType(of variable: Variable, at instrIndex: Int) -> Type {
+        if let instrTypes = types[variable],
+           let typeData = instrTypes.last(where: { el in el.index <= instrIndex })
+        {
+            return typeData.type
+        }
+
+        return .unknown
+    }
+
+    // Get type of variable at current instruction only if it is explicitely saved (type changes at this instruction)
+    // Used in adoption, so we do not adopt types which are not necessary
+    public func getSavedType(of variable: Variable, at instrIndex: Int) -> Type {
```

Comment:
Maybe call this `getTypeIfModified` or something like that? I think that'd be a bit more intuitive

Reply:
And I think this should return `nil` if the type was not modified

Reply:
hmm I don't really like having two nils (nil & .unknown).
But there is other way around so we don't need this function at all.

When beginning adoption, we do not append to `typeMaps` this struct but rather inverted one [index] -> [(Var, Type)]
Then it will be easy to adopt just runtime types that were present in original program

Reply:
With `nil` you can distinguish in the (probably rare?) case that the type is updated at the instruction but is actually set to .unknown (I guess that can only ever happen at the definition of a variable). But yeah, your suggestion sounds better, I agree :)